src/osd/PG.cc

```diff
@@ -574,9 +574,13 @@ bool PG::search_for_missing(
 bool PG::MissingLoc::readable_with_acting(
   const hobject_t &hoid,
   const set<pg_shard_t> &acting) const {
-  if (!needs_recovery(hoid)) return true;
+  if (!needs_recovery(hoid))
+    return true;
+  if (is_deleted(hoid))
+    return false;
```

Comment:
I figured the conservative path would be simpler here - treating deletes more like other writes, and not requiring any additional blocking or shenanigans with ObjectContext in each path that tries to use this object.

Like you said it just blocks reads on objects that were deleted until the delete occurs, which should not be a major part of many workloads, and didn't seem worth optimizing for. This is less strict than the current code, which would end up blocking all i/o for all deletes due to being part of peering.